contrib/src/main/java/com/datatorrent/contrib/kafka/POJOKafkaOutputOperator.java

```diff
@@ -0,0 +1,272 @@
+  private long messageCount;
+  private long byteCount;
+  private String brokerList;
+  private double windowTimeSec;
+  private String keyField = "";
+  protected boolean isBatchProcessing = true;
+  @Min(2)
+  protected int batchSize = DEFAULT_BATCH_SIZE;
+  protected transient List<KeyedMessage<Object,Object>> messageList = new ArrayList<>();
+  protected transient PojoUtils.Getter keyMethod;
+  protected transient Class<?> pojoClass;
+  protected transient ProducerConfig producerConfig;
+
+  public final transient DefaultInputPort<Object> inputPort = new DefaultInputPort<Object>() {
+    @Override
+    public void setup(Context.PortContext context)
+    {
+      if(context.getAttributes().contains(Context.PortContext.TUPLE_CLASS)) {
+        pojoClass = context.getAttributes().get(Context.PortContext.TUPLE_CLASS);
+      }
+    }
+
+    @Override
+    public void process(Object tuple)
+    {
+      processTuple(tuple);
+    }
+  };
+
+  /**
+   * setup producer configuration.
+   * @return ProducerConfig
+   */
+  @Override
+  protected ProducerConfig createKafkaProducerConfig(){
+    if(brokerList != null) {
+      getConfigProperties().setProperty(BROKER_KEY, brokerList);
+    }
+    getConfigProperties().setProperty(BATCH_NUM_KEY, String.valueOf(batchSize));
+    producerConfig = super.createKafkaProducerConfig();
+    return producerConfig;
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    windowTimeSec = (context.getValue(Context.OperatorContext.APPLICATION_WINDOW_COUNT)
+      * context.getValue(Context.DAGContext.STREAMING_WINDOW_SIZE_MILLIS) * 1.0) / 1000.0;
+    if(pojoClass != null && keyField != "") {
+      try {
+        keyMethod = generateGetterForKeyField();
+      } catch (NoSuchFieldException e) {
+        throw new RuntimeException("Field " + keyField + " is invalid: " + e);
+      }
+    }
+    batchSize = producerConfig.batchNumMessages();
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    super.beginWindow(windowId);
+    outputMessagesPerSec = 0;
+    outputBytesPerSec = 0;
+    messageCount = 0;
+    byteCount = 0;
+  }
+
+  /**
+   * Write the incoming tuple to Kafka
+   * @param tuple incoming tuple
+   */
+  protected void processTuple(Object tuple)
+  {
+    // Get the getter method from the keyField
+    if(keyMethod == null && keyField != "") {
+      pojoClass = tuple.getClass();
+      try {
+        keyMethod = generateGetterForKeyField();
+      } catch (NoSuchFieldException e) {
+        throw new RuntimeException("Field " + keyField + " is invalid: " + e);
+      }
+    }
+
+    // Convert the given tuple to KeyedMessage
+    KeyedMessage msg;
+    if(keyMethod != null) {
+      msg = new KeyedMessage(getTopic(), keyMethod.get(tuple), tuple);
+    } else {
+      msg = new KeyedMessage(getTopic(), tuple, tuple);
+    }
+
+    if (isBatchProcessing) {
+      if (messageList.size() == batchSize) {
```

Comment:
I think you don't have to do batch manually. you can set the  batch.num.messages to batch size and also set the queue.buffering.max.ms to streaming window size. It's not strictly same as what you are doing here. But the logic and performance are similar